### PR TITLE
Fix hang in WriteToTable.

### PR DESCRIPTION
**Pull request by gtopper**

### CODE REVIEW

File: storey/flow.py [hunk 1]
```diff
@@ -532,18 +532,19 @@ def __init__(self, max_in_flight=8, **kwargs):
...
                         if pending_event.pending and not pending_event.in_flight:
-                            resp = await self._process_event(pending_event.pending)
+                            resp = await self._safe_process_events(pending_event.pending)
                             for event in pending_event.pending:
                                 await self._handle_completed(event, resp)
                     if self._q.empty():
                         break
                     else:
-                        await self._q.put(_termination_obj)
+                        job = _termination_obj
```
Comment (talIguaz) [inline storey/flow.py]: `job` is already `termination_obj`, no?

File: storey/flow.py [hunk 2]
```diff
@@ -559,11 +560,12 @@ async def _worker(self):
                     self._pending_by_key[event.key].pending = []
 
                     task = self._safe_process_events(self._pending_by_key[event.key].in_flight)
-                    await self._q.put((event, asyncio.get_running_loop().create_task(task)))
+                    job = (event, asyncio.get_running_loop().create_task(task))
```
Comment (talIguaz) [inline storey/flow.py]: Well what you are doing here is just prioritizing a specific key over the others.
If all keys keep coming all the time, you end up only processing `key1` first,  then `key2` etc

File: storey/flow.py (hunk 1 above)
Comment (gtopper) [reply storey/flow.py]: Hm yeah, maybe this is not needed? Why would q not be empty?

Comment (gtopper) [reply storey/flow.py]: I replaced the if/else with just a break.

File: storey/flow.py (hunk 2 above)
Comment (gtopper) [reply storey/flow.py]: Well, a bit... The producer will still block on the queue if there are more keys than queue slots. So there's still back pressure. Do we care about the exact prioritization of keys when writing to a table?

File: storey/flow.py [hunk 3]
```diff
@@ -531,20 +531,29 @@ def __init__(self, max_in_flight=8, **kwargs):
...
+                jobs = self_sent_jobs.pop(received_job_count, None)
+                if jobs:
+                    job = jobs[0]
+                    if len(jobs) > 1:
+                        self_sent_jobs[received_job_count] = jobs[1:]
+                else:
+                    job = await self._q.get()
+                    received_job_count += 1
+                    if job is _termination_obj:
+                        if received_job_count in self_sent_jobs:
```
Comment (talIguaz) [inline storey/flow.py:548]: Why would `received_job_count in self_sent_jobs` ? you are checking it at L539 and nothing else but this function isn’t adding to `self_sent_jobs` so it should not exist there, no?

File: storey/flow.py [hunk 4]
```diff
@@ -559,11 +568,14 @@ async def _worker(self):
                     self._pending_by_key[event.key].pending = []
 
                     task = self._safe_process_events(self._pending_by_key[event.key].in_flight)
-                    await self._q.put((event, asyncio.get_running_loop().create_task(task)))
+                    tail_position = received_job_count + self._q.qsize()
+                    jobs_at_tail = self_sent_jobs.get(tail_position, [])
```
Comment (talIguaz) [inline storey/flow.py:572]: Aren't you always gonna miss the get and init with []?

**Review by talIguaz — APPROVED** [review]

**Pull request merged**